core/templates/dev/head/pages/exploration_player/AudioPreloaderService.js

```diff
@@ -20,117 +20,124 @@ oppia.factory('AudioPreloaderService', [
+        ).then(function(loadedAudio) {
+          for (var i = 0; i < _filesBeingDownloaded.length; i++) {
+            if (_filesBeingDownloaded[i] === audioFile) {
+              _filesBeingDownloaded.splice(i, 1);
+              break;
+            }
+          }
+          if (_filesToBeDownloaded.length > 0) {
+            var audioFile = _filesToBeDownloaded.shift();
+            _filesBeingDownloaded.push(audioFile);
+            loadAudio(audioFile);
+          }
+          if (_audioLoadedCallback) {
+            _audioLoadedCallback(loadedAudio.filename);
+          }
+        });
+    };
 
-            $scope.cancel = function() {
-              $modalInstance.dismiss('cancel');
-            };
+    var _kickOffAudioPreloader = function(sourceStateName) {
+      _isRunning = true;
+      _filesToBeDownloaded = _getAudioFilesInBfsOrder(sourceStateName);
+      while (_filesBeingDownloaded.length <
+          MAX_NUM_AUDIO_FILES_TO_DOWNLOAD_SIMULTANEOUSLY && 
+          _filesToBeDownloaded.length > 0) {
+        var audioFile = _filesToBeDownloaded.shift();
+        _filesBeingDownloaded.push(audioFile);
+        loadAudio(audioFile);
+      }
+    };
 
-            $scope.chooseDifferentLanguage = function() {
-              $modalInstance.close({
-                shouldDownloadAllAudioInExploration: false,
-                shouldOpenSettingsModal: true
-              });
-            };
-          }]
-      }).result.then(function(result) {
-        if (result.shouldOpenSettingsModal) {
-          // If the user elected to choose a different language, open
-          // the settings modal (later can isolate to a language-only
-          // modal), and on the callback re-open the bandwidth confirmation
-          // modal if the file for the new language hasn't been loaded.
-          AudioTranslationManagerService
-            .showAudioTranslationSettingsModal(function(newLanguageCode) {
-              if (LanguageUtilService.isAutogeneratedAudioLanguage(
-                  newLanguageCode)) {
-                confirmationCallback(newLanguageCode);
-              } else {
-                var newAudioTranslation =
-                  audioTranslationsForContent[newLanguageCode];
-                if (newAudioTranslation && !AssetsBackendApiService.isCached(
-                  newAudioTranslation.filename)) {
-                  _showBandwidthConfirmationModal(
-                    audioTranslationsForContent, newLanguageCode,
-                    confirmationCallback)
-                }
-              }
-            });
-        } else {
-          confirmationCallback(languageCode);
-          if (result.shouldDownloadAllAudioInExploration) {
-            _preloadAllAudioFiles(languageCode);
-          }
-        }
+    var _cancelPreloading = function() {
+      _isRunning = false;
+      AssetsBackendApiService.abortAllCurrentDownloads();
+      _filesBeingDownloaded.forEach(function(file) {
+        _filesToBeDownloaded.push(file);
```

Comment:
In the subsequent kickoff, it looks like this list just gets wiped. Is that the intent? If so, should _cancelPreloading() just clear the lists, or is there a reason I'm missing for retaining the info?

We should file an issue to look into pausing, btw. Otherwise it's possible that we try to download an audio file multiple times and incur corresponding bandwidth costs.

Reply:
You're right, I removed it.

Making a note to file an issue to look into pausing, I'll make it after this PR gets in